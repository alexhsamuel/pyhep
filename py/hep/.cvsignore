*.pyc
*.pyo
config.py
*-install
pdt.pickle
