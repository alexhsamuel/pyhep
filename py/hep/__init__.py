#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


