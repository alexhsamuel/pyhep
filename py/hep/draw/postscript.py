#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import division

import fonts
from   hep.draw import *
import re

#-----------------------------------------------------------------------
# constants
#-----------------------------------------------------------------------

# Map from font names to PostScript font family names.
font_map = {
    "Courier":                  "Courier",
    "Courier-Bold":             "Courier-Bold",
    "Courier-Oblique":          "Courier-Oblique",
    "Helvetica":                "Helvetica",
    "Helvetica-Bold":           "Helvetica-Bold",
    "Helvetica-Oblique":        "Helvetica-Oblique",
    "Symbol":                   "Symbol",
    "Symbol-Italic":            "Symbol",
    "Times":                    "Times-Roman",
    "Times-Bold":               "Times-Bold",
    "Times-Italic":             "Times-Italic",
    }


#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def drawMarker(shape, x, y, size, f):
    """Draw a marker.

    Draw marker 'shape' at '(x, y)' with 'size' by writing PostScript to
    file 'f'."""

    size /= 2
    if shape == "filled dot":
        f.write("newpath ")
        f.write("%f %f %f 0 360 arc " % (x, y, size))
        f.write("closepath fill\n")
    elif shape == "empty dot":
        f.write("newpath ")
        f.write("%f %f %f 0 360 arc " % (x, y, size))
        f.write("closepath stroke\n")
    elif shape == "filled square":
        size /= sqrt(2)
        f.write("newpath ")
        f.write("%f %f moveto " % (x + size, y + size))
        f.write("%f %f lineto " % (x - size, y + size))
        f.write("%f %f lineto " % (x - size, y - size))
        f.write("%f %f lineto " % (x + size, y - size))
        f.write("closepath fill\n")
    elif shape == "empty square":
        size /= sqrt(2)
        f.write("% empty square\n")
        f.write("newpath ")
        f.write("%f %f moveto " % (x + size, y + size))
        f.write("%f %f lineto " % (x - size, y + size))
        f.write("%f %f lineto " % (x - size, y - size))
        f.write("%f %f lineto " % (x + size, y - size))
        f.write("closepath stroke\n")
    elif shape == "filled diamond":
        f.write("newpath ")
        f.write("%f %f moveto " % (x + size, y))
        f.write("%f %f lineto " % (x, y + size))
        f.write("%f %f lineto " % (x - size, y))
        f.write("%f %f lineto " % (x, y - size))
        f.write("closepath fill\n")
    elif shape == "empty diamond":
        f.write("newpath ")
        f.write("%f %f moveto " % (x + size, y))
        f.write("%f %f lineto " % (x, y + size))
        f.write("%f %f lineto " % (x - size, y))
        f.write("%f %f lineto " % (x, y - size))
        f.write("closepath stroke\n")
    elif shape == "*":
        dx = 0.866025 * size
        dy = 0.5 * size
        f.write("newpath %f %f moveto " % (x,      y - size))
        f.write("%f %f lineto stroke "  % (x,      y + size))
        f.write("newpath %f %f moveto " % (x - dx, y - dy))
        f.write("%f %f lineto stroke "  % (x + dx, y + dy))
        f.write("newpath %f %f moveto " % (x - dx, y + dy))
        f.write("%f %f lineto stroke\n" % (x + dx, y - dy))
    elif shape == "+":
        f.write("newpath %f %f moveto " % (x - size, y))
        f.write("%f %f lineto stroke "  % (x + size, y))
        f.write("newpath %f %f moveto " % (x,        y - size))
        f.write("%f %f lineto stroke\n" % (x,        y + size))
    elif shape == "X":
        d = 0.707107 * size
        f.write("newpath %f %f moveto " % (x - d,  y - d))
        f.write("%f %f lineto stroke "  % (x + d,  y + d))
        f.write("newpath %f %f moveto " % (x - d,  y + d))
        f.write("%f %f lineto stroke\n" % (x + d,  y - d))
    else:
        raise NotImplementedError, "marker shape '%s'" % shape


#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Renderer(object):

    def __init__(self, size, style):
        self.size = size
        self.style = style
        # These attributes note the current graphic state as we are
        # generating PostScript.
        self.current_scale = 1


    def _writeDefs(self):
        pass


    def _writePageSetup(self):
        self.file.write("%f %f scale\n" % (1 / point, 1 / point))
        self.file.write("true setstrokeadjust\n")
        # Clear out current graphics state so that it will get set again
        # on this page.
        self.__color = None
        self.__thickness = None
        self.__dash = None
        self.__font = None


    def _writePageCleanup(self):
        self.file.write("showpage\n")


    def line(self, points):
        x, y = points[0]
        self.file.write("newpath %.6f %.6f moveto " % (x, y))
        for x, y in points[1 :]:
            self.file.write("%.6f %.6f lineto " % (x, y))
        self.file.write("stroke\n")


    def polygon(self, points):
        x, y = points[0]
        self.file.write("newpath %f %f moveto " % (x, y))
        for x, y in points[1 :]:
            self.file.write("%f %f lineto " % (x, y))
        self.file.write("closepath fill\n")


    def text(self, position, text, alignment):
        # Don't map the position; text is drawn in physical units.
        x, y = position
        # Escape special characters.
        text = re.sub(r"([(\\)])", r"\\\1", text)
        # Align the text.
        x_alignment, y_alignment = alignment
        width, height, depth = fonts.getTextExtent(text, self.font)
        x -= x_alignment * width
        y -= y_alignment * height
        # Draw the text.
        self.file.write("%f %f moveto (%s) show\n" % (x, y, text))


    def symbol(self, position, symbol):
        # Don't map the position; text is drawn in physical units.
        x, y = position

        font_family, font_size = self.font
        # Determine whether the symbol should be roman or italic.
        italic = font_family.endswith("-Italic") \
                 or font_family.endswith("-Oblique")
        # Loop up the font family and PostScript glyph name for this
        # symbol. 
        symbol_family, glyph_name = fonts.symbol_map[symbol]
        if symbol_family is None:
            symbol_family = font_family
        else:
            if italic:
                symbol_family += "-Italic"
            self.font = (symbol_family, font_size)

        # Draw the symbol glyph.
        self.file.write(
            "%f %f moveto /%s glyphshow\n" % (x, y, glyph_name))


    def marker(self, position, shape, size):
        x, y = position
        drawMarker(shape, x, y, size, self.file)


    def __set_color(self, color):
        if color != self.__color:
            self.file.write(
                "%f %f %f setrgbcolor\n"
                % (color.red, color.green, color.blue))
            self.__color = color


    color = property(lambda self: self.__color, __set_color)


    def __set_thickness(self, thickness):
        if thickness != self.__thickness:
            self.file.write("%f setlinewidth\n" % thickness)
            self.__thickness = thickness


    thickness = property(lambda self: self.__tickness, __set_thickness)


    def __set_dash(self, dash):
        if dash != self.__dash:
            self.file.write("[")
            if dash is not None:
                self.file.write(" ".join(map(str, dash)))
            self.file.write("] 0 setdash\n")
            self.__dash = dash


    dash = property(lambda self: self.__dash, __set_dash)


    def __set_font(self, font_spec):
        if font_spec != self.__font:
            font_name, font_size = font_spec
            family = font_map.get(font_name, font_name)
            # Look up the font family in our font map, to determine the
            # PostScript font family name.
            self.file.write(
                "/%s findfont %f scalefont "
                % (family, font_size))
            # Since we don't have an italic symbol font, fake it by
            # applying a shearing transformation.
            if font_name == "Symbol-Italic":
                self.file.write("[ 1 0 0.167 1 0 0 ] makefont ")
            self.file.write("setfont\n")
            self.__font = font_spec

    font = property(lambda self: self.__font, __set_font)



#-----------------------------------------------------------------------

class PSFile(Renderer):
    """Renderer to a multi-page ADSC PostScript file.

    Each time 'begin' is called, a new page is created."""

    default_page_size = "letter"
    

    def __init__(self, path, page_size=None, **style):
        """Create a renderer to a PostScript file.

        'path' -- Path to the new PostScript file.

        'page_size' -- The page size, either a pair of lengths or the
        name of a standard page size.  Default the value of the class
        attribute 'default_page_size'."""

        # Use the default page size if unspecified.
        if page_size is None:
            page_size = self.default_page_size
        # Decode standard page size names.
        page_size = parsePageSize(page_size)

        # Initialize the base class.
        Renderer.__init__(self, page_size, style)
        # Page numbering starts at one.
        self.page = 1

        # Set up the file.
        self.file = file(path, "w")
        self.file.write("%!PS-Adobe-1.0\n")
        self.file.write("%%BoundingBox: " + "0 0 %d %d\n"
                        % (page_size[0] / point, page_size[1] / point))
        self.file.write("%%EndComments\n")
        self._writeDefs()
        

    def render(self, figure):
        # Initialize.
        self.file.write("%%%%Page: %d %d\n" % (self.page, self.page + 1))
        self._writePageSetup()
        # Render the figure.
        sx, sy = self.size
        region = hep.draw.adjustRegionForStyle((0, 0, sx, sy), self.style)
        figure._render(self, region)
        # Clean up.
        self._writePageCleanup()
        self.page += 1



#-----------------------------------------------------------------------

class EPSFile(Renderer):
    """A renderer to an encapsulated PostScript file.

    The 'begin' method should only be invoked once."""

    default_size = (0.08, 0.05)


    def __init__(self, path, size=None, **style):
        """Create an encapsulated PostScript renderer.

        'path' -- Path to the PostScript file.

        'size' -- The bounding box size of the figure."""

        # Use the default size if unspecified.
        if size is None:
            size = self.default_size

        self.__path = path
        self.rendered = False

        # Initialize the base class.
        Renderer.__init__(self, size, style)


    def render(self, figure):
        if self.rendered:
            raise RuntimeError, "an 'EPSFile' may only be rendered once"

        sx, sy = self.size

        # Set up the file.
        self.file = file(self.__path, "w")
        self.file.write("%!PS-Adobe-1.0 EPSF-1.0\n")
        self.file.write("%%%%BoundingBox: 0 0 %d %d\n"
                        % (sx / point, sy / point))
        self.file.write("%%EndComments\n")
        self._writeDefs()
        self._writePageSetup()
        # Render.
        region = hep.draw.adjustRegionForStyle((0, 0, sx, sy), self.style)
        figure._render(self, (0, 0, sx, sy))
        # Clean up.
        self._writePageCleanup()
        del self.file

        self.rendered = True


