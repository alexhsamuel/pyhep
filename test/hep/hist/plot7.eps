%!PS-Adobe-1.0 EPSF-1.0
%%BoundingBox: 0 0 651 283
%%EndComments
2834.645669 2834.645669 scale
true setstrokeadjust
0.000000 0.000000 0.000000 setrgbcolor
0.000176 setlinewidth
newpath 0.015000 0.022667 moveto 0.015000 0.021667 lineto stroke
newpath 0.026875 0.022667 moveto 0.026875 0.021667 lineto stroke
newpath 0.038750 0.022667 moveto 0.038750 0.021667 lineto stroke
newpath 0.050625 0.022667 moveto 0.050625 0.021667 lineto stroke
newpath 0.062500 0.022667 moveto 0.062500 0.021667 lineto stroke
newpath 0.074375 0.022667 moveto 0.074375 0.021667 lineto stroke
newpath 0.086250 0.022667 moveto 0.086250 0.021667 lineto stroke
newpath 0.098125 0.022667 moveto 0.098125 0.021667 lineto stroke
newpath 0.110000 0.022667 moveto 0.110000 0.021667 lineto stroke
/Times-Roman findfont 0.003175 scalefont setfont
0.013995 0.018127 moveto /hyphen glyphshow
0.015052 0.018127 moveto (6) show
0.016640 0.018127 moveto (.) show
0.017433 0.018127 moveto (0) show
0.024362 0.018127 moveto /hyphen glyphshow
0.025419 0.018127 moveto (4) show
0.027007 0.018127 moveto (.) show
0.027801 0.018127 moveto (0) show
0.036237 0.018127 moveto /hyphen glyphshow
0.037294 0.018127 moveto (2) show
0.038882 0.018127 moveto (.) show
0.039676 0.018127 moveto (0) show
0.048641 0.018127 moveto (0) show
0.050228 0.018127 moveto (.) show
0.051022 0.018127 moveto (0) show
0.060516 0.018127 moveto (2) show
0.062103 0.018127 moveto (.) show
0.062897 0.018127 moveto (0) show
0.072391 0.018127 moveto (4) show
0.073978 0.018127 moveto (.) show
0.074772 0.018127 moveto (0) show
0.084266 0.018127 moveto (6) show
0.085853 0.018127 moveto (.) show
0.086647 0.018127 moveto (0) show
0.096141 0.018127 moveto (8) show
0.097728 0.018127 moveto (.) show
0.098522 0.018127 moveto (0) show
0.105555 0.018127 moveto (1) show
0.107142 0.018127 moveto (0) show
0.108730 0.018127 moveto (.) show
0.109524 0.018127 moveto (0) show
newpath 0.014000 0.023667 moveto 0.013000 0.023667 lineto stroke
newpath 0.014000 0.037623 moveto 0.013000 0.037623 lineto stroke
newpath 0.014000 0.051580 moveto 0.013000 0.051580 lineto stroke
newpath 0.014000 0.065537 moveto 0.013000 0.065537 lineto stroke
newpath 0.014000 0.079493 moveto 0.013000 0.079493 lineto stroke
0.010412 0.022397 moveto (0) show
0.008825 0.036353 moveto (5) show
0.010412 0.036353 moveto (0) show
0.007237 0.050310 moveto (1) show
0.008825 0.050310 moveto (0) show
0.010412 0.050310 moveto (0) show
0.007237 0.064267 moveto (1) show
0.008825 0.064267 moveto (5) show
0.010412 0.064267 moveto (0) show
0.007237 0.078223 moveto (2) show
0.008825 0.078223 moveto (0) show
0.010412 0.078223 moveto (0) show
0.000000 0.082873 moveto (entries) show
0.008465 0.082873 moveto ( ) show
0.009258 0.082873 moveto (/) show
0.010141 0.082873 moveto ( ) show
0.010935 0.082873 moveto (0) show
0.012522 0.082873 moveto (.) show
0.013316 0.082873 moveto (5) show
0.014903 0.082873 moveto ( ) show
[0.0002 0.0002] 0 setdash
0.000035 setlinewidth
0.000176 setlinewidth
[] 0 setdash
newpath 0.015000 0.023667 moveto 0.110000 0.023667 lineto stroke
0.000282 setlinewidth
newpath 0.020938 0.023667 moveto 0.023906 0.023667 lineto stroke
newpath 0.022422 0.023667 moveto 0.022422 0.023987 lineto stroke
newpath 0.022422 0.023667 0.000750 0 360 arc closepath fill
newpath 0.023906 0.023667 moveto 0.026875 0.023667 lineto stroke
newpath 0.025391 0.023667 moveto 0.025391 0.023987 lineto stroke
newpath 0.025391 0.023667 0.000750 0 360 arc closepath fill
newpath 0.026875 0.023667 moveto 0.029844 0.023667 lineto stroke
newpath 0.028359 0.023667 moveto 0.028359 0.023987 lineto stroke
newpath 0.028359 0.023667 0.000750 0 360 arc closepath fill
newpath 0.029844 0.023667 moveto 0.032813 0.023667 lineto stroke
newpath 0.031328 0.023667 moveto 0.031328 0.023987 lineto stroke
newpath 0.031328 0.023667 0.000750 0 360 arc closepath fill
newpath 0.032813 0.025621 moveto 0.035781 0.025621 lineto stroke
newpath 0.034297 0.024822 moveto 0.034297 0.026349 lineto stroke
newpath 0.034297 0.025621 0.000750 0 360 arc closepath fill
newpath 0.035781 0.028133 moveto 0.038750 0.028133 lineto stroke
newpath 0.037266 0.026971 moveto 0.037266 0.029234 lineto stroke
newpath 0.037266 0.028133 0.000750 0 360 arc closepath fill
newpath 0.038750 0.039577 moveto 0.041719 0.039577 lineto stroke
newpath 0.040234 0.037434 moveto 0.040234 0.041664 lineto stroke
newpath 0.040234 0.039577 0.000750 0 360 arc closepath fill
newpath 0.041719 0.044322 moveto 0.044688 0.044322 lineto stroke
newpath 0.043203 0.041886 moveto 0.043203 0.046702 lineto stroke
newpath 0.043203 0.044322 0.000750 0 360 arc closepath fill
newpath 0.044688 0.068049 moveto 0.047656 0.068049 lineto stroke
newpath 0.046172 0.064529 moveto 0.046172 0.071568 lineto stroke
newpath 0.046172 0.068049 0.000750 0 360 arc closepath fill
newpath 0.047656 0.075864 moveto 0.050625 0.075864 lineto stroke
newpath 0.049141 0.072047 moveto 0.049141 0.079682 lineto stroke
newpath 0.049141 0.075864 0.000750 0 360 arc closepath fill
newpath 0.050625 0.073352 moveto 0.053594 0.073352 lineto stroke
newpath 0.052109 0.069628 moveto 0.052109 0.077076 lineto stroke
newpath 0.052109 0.073352 0.000750 0 360 arc closepath fill
newpath 0.053594 0.072515 moveto 0.056562 0.072515 lineto stroke
newpath 0.055078 0.068822 moveto 0.055078 0.076207 lineto stroke
newpath 0.055078 0.072515 0.000750 0 360 arc closepath fill
newpath 0.056562 0.047393 moveto 0.059531 0.047393 lineto stroke
newpath 0.058047 0.044785 moveto 0.058047 0.049945 lineto stroke
newpath 0.058047 0.047393 0.000750 0 360 arc closepath fill
newpath 0.059531 0.035111 moveto 0.062500 0.035111 lineto stroke
newpath 0.061016 0.033286 moveto 0.061016 0.036879 lineto stroke
newpath 0.061016 0.035111 0.000750 0 360 arc closepath fill
newpath 0.062500 0.027854 moveto 0.065469 0.027854 lineto stroke
newpath 0.063984 0.026726 moveto 0.063984 0.028920 lineto stroke
newpath 0.063984 0.027854 0.000750 0 360 arc closepath fill
newpath 0.065469 0.024783 moveto 0.068437 0.024783 lineto stroke
newpath 0.066953 0.024142 moveto 0.066953 0.025336 lineto stroke
newpath 0.066953 0.024783 0.000750 0 360 arc closepath fill
newpath 0.068437 0.024225 moveto 0.071406 0.024225 lineto stroke
newpath 0.069922 0.023667 moveto 0.069922 0.024649 lineto stroke
newpath 0.069922 0.024225 0.000750 0 360 arc closepath fill
newpath 0.071406 0.023667 moveto 0.074375 0.023667 lineto stroke
newpath 0.072891 0.023667 moveto 0.072891 0.023987 lineto stroke
newpath 0.072891 0.023667 0.000750 0 360 arc closepath fill
newpath 0.074375 0.023667 moveto 0.077344 0.023667 lineto stroke
newpath 0.075859 0.023667 moveto 0.075859 0.023987 lineto stroke
newpath 0.075859 0.023667 0.000750 0 360 arc closepath fill
newpath 0.077344 0.023667 moveto 0.080312 0.023667 lineto stroke
newpath 0.078828 0.023667 moveto 0.078828 0.023987 lineto stroke
newpath 0.078828 0.023667 0.000750 0 360 arc closepath fill
0.600000 0.000000 0.000000 setrgbcolor
newpath 0.015000 0.024016 moveto 0.019750 0.024016 lineto stroke
newpath 0.017375 0.023667 moveto 0.017375 0.024281 lineto stroke
newpath 0.017375 0.024016 0.000750 0 360 arc closepath fill
newpath 0.019750 0.023841 moveto 0.024500 0.023841 lineto stroke
newpath 0.022125 0.023667 moveto 0.022125 0.024078 lineto stroke
newpath 0.022125 0.023841 0.000750 0 360 arc closepath fill
newpath 0.024500 0.024190 moveto 0.029250 0.024190 lineto stroke
newpath 0.026875 0.023816 moveto 0.026875 0.024491 lineto stroke
newpath 0.026875 0.024190 0.000750 0 360 arc closepath fill
newpath 0.029250 0.024364 moveto 0.034000 0.024364 lineto stroke
newpath 0.031625 0.023964 moveto 0.031625 0.024710 lineto stroke
newpath 0.031625 0.024364 0.000750 0 360 arc closepath fill
newpath 0.034000 0.026632 moveto 0.038750 0.026632 lineto stroke
newpath 0.036375 0.025885 moveto 0.036375 0.027342 lineto stroke
newpath 0.036375 0.026632 0.000750 0 360 arc closepath fill
newpath 0.038750 0.030645 moveto 0.043500 0.030645 lineto stroke
newpath 0.041125 0.029518 moveto 0.041125 0.031736 lineto stroke
newpath 0.041125 0.030645 0.000750 0 360 arc closepath fill
newpath 0.043500 0.033262 moveto 0.048250 0.033262 lineto stroke
newpath 0.045875 0.031945 moveto 0.045875 0.034543 lineto stroke
newpath 0.045875 0.033262 0.000750 0 360 arc closepath fill
newpath 0.048250 0.040589 moveto 0.053000 0.040589 lineto stroke
newpath 0.050625 0.038850 moveto 0.050625 0.042293 lineto stroke
newpath 0.050625 0.040589 0.000750 0 360 arc closepath fill
newpath 0.053000 0.048963 moveto 0.057750 0.048963 lineto stroke
newpath 0.055375 0.046862 moveto 0.055375 0.051064 lineto stroke
newpath 0.055375 0.048963 0.000750 0 360 arc closepath fill
newpath 0.057750 0.048265 moveto 0.062500 0.048265 lineto stroke
newpath 0.060125 0.046194 moveto 0.060125 0.050337 lineto stroke
newpath 0.060125 0.048265 0.000750 0 360 arc closepath fill
newpath 0.062500 0.054197 moveto 0.067250 0.054197 lineto stroke
newpath 0.064875 0.051889 moveto 0.064875 0.056505 lineto stroke
newpath 0.064875 0.054197 0.000750 0 360 arc closepath fill
newpath 0.067250 0.044602 moveto 0.072000 0.044602 lineto stroke
newpath 0.069625 0.042691 moveto 0.069625 0.046513 lineto stroke
newpath 0.069625 0.044602 0.000750 0 360 arc closepath fill
newpath 0.072000 0.040066 moveto 0.076750 0.040066 lineto stroke
newpath 0.074375 0.038353 moveto 0.074375 0.041743 lineto stroke
newpath 0.074375 0.040066 0.000750 0 360 arc closepath fill
newpath 0.076750 0.035181 moveto 0.081500 0.035181 lineto stroke
newpath 0.079125 0.033741 moveto 0.079125 0.036585 lineto stroke
newpath 0.079125 0.035181 0.000750 0 360 arc closepath fill
newpath 0.081500 0.027505 moveto 0.086250 0.027505 lineto stroke
newpath 0.083875 0.026660 moveto 0.083875 0.028312 lineto stroke
newpath 0.083875 0.027505 0.000750 0 360 arc closepath fill
newpath 0.086250 0.025760 moveto 0.091000 0.025760 lineto stroke
newpath 0.088625 0.025125 moveto 0.088625 0.026356 lineto stroke
newpath 0.088625 0.025760 0.000750 0 360 arc closepath fill
newpath 0.091000 0.024364 moveto 0.095750 0.024364 lineto stroke
newpath 0.093375 0.023964 moveto 0.093375 0.024710 lineto stroke
newpath 0.093375 0.024364 0.000750 0 360 arc closepath fill
newpath 0.095750 0.024016 moveto 0.100500 0.024016 lineto stroke
newpath 0.098125 0.023667 moveto 0.098125 0.024281 lineto stroke
newpath 0.098125 0.024016 0.000750 0 360 arc closepath fill
newpath 0.100500 0.023667 moveto 0.105250 0.023667 lineto stroke
newpath 0.102875 0.023667 moveto 0.102875 0.023867 lineto stroke
newpath 0.102875 0.023667 0.000750 0 360 arc closepath fill
newpath 0.105250 0.023667 moveto 0.110000 0.023667 lineto stroke
newpath 0.107625 0.023667 moveto 0.107625 0.023867 lineto stroke
newpath 0.107625 0.023667 0.000750 0 360 arc closepath fill
