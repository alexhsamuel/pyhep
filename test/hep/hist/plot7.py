"""Test plotting of histogram statistics."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.draw import *
import hep.draw.postscript
import hep.draw.xwindow
import hep.hist
from   hep.hist.plot import Plot, Statistics
import sys

#-----------------------------------------------------------------------
# tests
#-----------------------------------------------------------------------

green = Color(0.2, 0.6, 0.4)

histogram1 = hep.hist.makeSample1D("normal gaussian")
histogram2 = hep.hist.makeSample1D("gaussian")

layout = GridLayout(2, 1)
plot = Plot(1, overflows=False,
            marker="filled dot", marker_size=0.0015)
plot.append(histogram1, color=black)
plot.append(histogram2, color=Color(0.6, 0, 0))
plot.append(Statistics(
    ("sum", "mean", "sd"), histogram1, histogram2))
layout[0, 0] = plot

plot = Plot(1, overflows=False,
            marker="filled dot", marker_size=0.0015)
plot.append(histogram1, color=black)
plot.append(histogram2, color=green)
plot.append(Statistics(
    ("sum", "overflows"), histogram2,
    annotation_color=green, annotation_position="left"))
layout[1, 0] = plot

hep.draw.postscript.EPSFile("plot7.eps", (0.23, 0.1)).render(layout)
window = hep.draw.xwindow.FigureWindow(layout, (0.23, 0.1))

if len(sys.argv) > 1:
    raw_input("hit enter to end: ")
