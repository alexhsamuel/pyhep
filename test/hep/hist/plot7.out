Traceback (most recent call last):
  File "plot7.py", line 41, in ?
    hep.draw.postscript.EPSFile("plot7.eps", (0.23, 0.1)).render(layout)
  File "/home/samuel/dev/pyhep/py/hep/draw/postscript.py", line 345, in render
    figure._render(self, (0, 0, sx, sy))
  File "/home/samuel/dev/pyhep/py/hep/draw/__init__.py", line 380, in _render
    figure._render(renderer, (x0, y0, x1, y1))
  File "/home/samuel/dev/pyhep/py/hep/hist/plot.py", line 1578, in _render
    s._render(
  File "/home/samuel/dev/pyhep/py/hep/hist/plot.py", line 1265, in _render
    self << formatStatistics(self.statistics, histogram)
  File "/home/samuel/dev/pyhep/py/hep/hist/plot.py", line 2119, in formatStatistics
    raise ValueError, "unknown statistic %r" % stat
ValueError: unknown statistic 'sd'
